### PR TITLE
Fix depwarn and error on 0.6

### PR DESCRIPTION
**Pull request by yuyichao**
Together with a lot of cleanup including but not limited to.

* Remove a few broken and unused macros
* Fix test when more than one platform is available
* Use `Ref` instead of single element/zero dimensional array

Travis/push seems fine though hello world test is failing for me on Nvidia and one of the context creation test is failing for me on Beignet...

### CODE REVIEW

**Comment by yuyichao** [comment]
..... The Beignet failure is caused by it being not happy about a NULL callback an a non-NULL userdata. The ccall wrapper use `Any` to pass the userdata so it's never `NULL` even when `C_NULL` is the parameter. I'll make it pass `Ptr` through directly and only use object address for other objects.

**Comment by yuyichao** [comment]
userdata failure on Beignet fixed. Not sure about the hello world one..... The string is `"\0\0\0hello wo"` with the nvidia driver....

**Comment by vchuravy** [comment]
What NVIDIA driver version are you using? Test pass for me on `378.13` I am running into a test failure with the Intel SDK 2016 R2 but those tests pass on NVIDIA and AMD CPU.

```

Intel(R) OpenCL, Intel(R) Xeon(R) CPU E5-2643 v3 @ 3.40GHz: Error During Test
  Test threw an exception of type OpenCL.cl.CLError
  Expression: cl.work_group_info(k, sf, device) != nothing
  CLError(code=-30, CL_INVALID_VALUE)
  Stacktrace:
   [1] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/src/macros.jl:6 [inlined]
   [2] work_group_info(::OpenCL.cl.Kernel, ::UInt32, ::OpenCL.cl.Device) at /home/wallnuss/.julia/v0.6/OpenCL/src/kernel.jl:121
   [3] work_group_info(::OpenCL.cl.Kernel, ::Symbol, ::OpenCL.cl.Device) at /home/wallnuss/.julia/v0.6/OpenCL/src/kernel.jl:136
   [4] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:66 [inlined]
   [5] macro expansion at ./test.jl:914 [inlined]
   [6] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:0 [inlined]
   [7] macro expansion at ./test.jl:853 [inlined]
   [8] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:3 [inlined]
   [9] macro expansion at ./test.jl:853 [inlined]
   [10] anonymous at ./<missing>:?
Intel(R) OpenCL, Intel(R) Xeon(R) CPU E5-2643 v3 @ 3.40GHz: Error During Test
  Test threw an exception of type OpenCL.cl.CLError
  Expression: cl.work_group_info(k, clf, device) != nothing
  CLError(code=-30, CL_INVALID_VALUE)
  Stacktrace:
   [1] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/src/macros.jl:6 [inlined]
   [2] work_group_info(::OpenCL.cl.Kernel, ::UInt32, ::OpenCL.cl.Device) at /home/wallnuss/.julia/v0.6/OpenCL/src/kernel.jl:121
   [3] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:67 [inlined]
   [4] macro expansion at ./test.jl:914 [inlined]
   [5] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:0 [inlined]
   [6] macro expansion at ./test.jl:853 [inlined]
   [7] macro expansion at /home/wallnuss/.julia/v0.6/OpenCL/test/test_kernel.jl:3 [inlined]
   [8] macro expansion at ./test.jl:853 [inlined]
   [9] anonymous at ./<missing>:?
```

**Comment by yuyichao** [comment]
I'm on nvidia 378.13 too. Hardware is `Quadro M1200`.

It really looks like a compiler bug to me since assigning `hw[0]` and `idx` to the output are both fine but assigning `hw[idx]` seems to have an offset....

**Comment by vchuravy** [comment]
My hardware is a `M6000`  what cuda version do you have installed?

**Comment by yuyichao** [comment]
`8.0.44`

**Comment by vchuravy** [comment]
Ah I can reproduce the error on NVIDIA now after #131

File: test/test_platform.jl [hunk 1]
```diff
@@ -23,9 +23,9 @@
 
         if length(cl.platforms()) > 1
             for p in cl.platforms()[2:end]
-                @test pointer(platform) == pointer(p)
-                @test hash(platform) == hash(p)
-                @test isequal(platform, p)
+                @test pointer(platform) != pointer(p)
+                @test hash(platform) != hash(p)
+                @test !isequal(platform, p)
```
Comment (tkelman) [inline test/test_platform.jl]: was this just never triggering before?

Comment (yuyichao) [inline test/test_platform.jl]: https://github.com/JuliaGPU/OpenCL.jl/issues/126

**Pull request merged**